async.el

```diff
@@ -197,11 +197,29 @@ It is intended to be used as follows:
                              (process-name proc) (process-exit-status proc))))
           (set (make-local-variable 'async-callback-value-set) t))))))
 
-(defun async-read-from-client (proc string)
-  ;; log the message in the process buffer
-  (with-current-buffer (process-buffer proc)
-    (insert string))
+(defun async--process-filter-line-buffering-decorator (filter)
+  "Decorate a process FILTER with line-buffering logic.
+
+Return a new process filter based on FILTER.
+
+FILTER is a function which could be used with `set-process-filter'.
 
+This decorator buffers input until it can pass a complete line
+further to the supplied FILTER.  It is useful as a buffer between
+a process producing data and an Emacs function operating on the
+data which expects to get complete lines as input."
+  (let ((data ""))
+    (lambda (process string)
+      (with-current-buffer (process-buffer process)
+        (insert string))
+
+      (let* ((line-data (split-string (concat data string) "\n")))
+        (while (cdr line-data)
+          (funcall filter process (car line-data))
```

Comment:
This means `filter` will never see a newline, right? That doesn't seem right.

Reply:
This might possibly be an issue if some other thing in async would rely on the newline.  But the return value code does not and the "message" code also just reads a sexp which is one-line (base64) so it should be fine.  I guess we can add a newline to the end of the data though.

Should I make a patch?